### PR TITLE
ARROW-12031: [C++][CSV] infer CSV timestamps columns with fractional seconds

### PR DESCRIPTION
**Pull request by MartinNowak**
- implement support for inferring CSV timestamps columns with fractional seconds
- use timestamp[ns] parser which also covers [ms] and [us] to
  avoid testing 2 additional timestamp types

### CODE REVIEW

**Review by pitrou — COMMENTED** [review]
Thank you very much. How about adding also a Python test somewhere near https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_csv.py#L510 ?

**Comment by MartinNowak** [comment]
> Thank you very much. How about adding also a Python test somewhere near https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_csv.py#L510 ?

OK, would rely on the CI to test/debug this so that I don't have to bootstrap another build/test system, hence didn't include this in the first run.

**Comment by pitrou** [comment]
Thanks @MartinNowak . You can use [Archery](https://arrow.apache.org/docs/developers/archery.html) (or run docker-compose directly) if you don't want setup a build environment yourself.

For example, to see the list of possible Python build images:
```bash
$ archery docker images | grep python
 - conda-python
 - conda-python-dask
 - conda-python-hdfs
 - conda-python-jpype
 - conda-python-kartothek
 - conda-python-pandas
 - conda-python-spark
 - conda-python-turbodbc
 - debian-python
 - fedora-python
 - python-sdist
 - python-wheel-manylinux-2010
 - python-wheel-manylinux-2014
 - python-wheel-manylinux-test-imports
 - python-wheel-manylinux-test-unittests
 - python-wheel-windows-test
 - python-wheel-windows-vs2017
 - ubuntu-cuda-python
 - ubuntu-python
 - ubuntu-python-sdist-test
```
and then run one of those images using `docker run <image-name>`.

However, since this PR is simple, I can probably fix issues directly if you have trouble building locally.

**Review by pitrou — APPROVED** [review]
+1, thank you. Will wait for CI before merging.

**Comment by MartinNowak** [comment]
> +1, thank you. Will wait for CI before merging.

Was actually quite smooth to build and test the python package. Thanks for the [good docs](https://github.com/apache/arrow/blob/master/docs/source/developers/python.rst#unit-testing) and well structured project.

**Pull request closed without merging**
